source/extensions/quic_listeners/quiche/envoy_quic_alarm.cc

```diff
@@ -0,0 +1,30 @@
+#include "extensions/quic_listeners/quiche/envoy_quic_alarm.h"
+
+namespace Envoy {
+namespace Quic {
+
+EnvoyQuicAlarm::EnvoyQuicAlarm(Event::Scheduler& scheduler, quic::QuicClock& clock,
+                               quic::QuicArenaScopedPtr<quic::QuicAlarm::Delegate> delegate)
+    : QuicAlarm(std::move(delegate)), scheduler_(scheduler), clock_(clock) {
+  timer_ = scheduler_.createTimer([this]() { Fire(); });
+}
+
+void EnvoyQuicAlarm::CancelImpl() { timer_->disableTimer(); }
+
+void EnvoyQuicAlarm::SetImpl() {
+  timer_->enableTimerInUs(std::chrono::microseconds(getDurationBeforeDeadline().ToMicroseconds()));
+}
+
+void EnvoyQuicAlarm::UpdateImpl() { SetImpl(); }
+
+quic::QuicTime::Delta EnvoyQuicAlarm::getDurationBeforeDeadline() {
+  quic::QuicTime::Delta duration(quic::QuicTime::Delta::Zero());
+  quic::QuicTime now = clock_.ApproximateNow();
+  if (deadline() > now) {
```

Comment:
i would capture deadline() in a temp rather than calling it twice.

Reply:
Done. But deadline is just a non-virtual getter actually.

Reply:
FWIW I have seen (and caused) things that previously were simple observers into much less trivial operations. We are working through one such issue now because someone (me :) changed Counter::name() from simply returning a reference to an owned string to taking a lock on a symbol table, decoding a StatName into an array of symbols, and joining the symbols together.

In Envoy style, naming something `Foo foo() const` does not imply its just returning a member variable; it may be doing heavy computation (or changed in the future to do so).

So in general I feel it's safer to captures such things in temps rather than "compute" them twice.